storage/replica_command.go

```diff
@@ -524,35 +520,68 @@ func (r *Replica) RangeLookup(batch engine.Engine, args proto.RangeLookupRequest
 		rangeCount = 1
 	}
 
-	var kvs []proto.KeyValue
+	var kvs []proto.KeyValue        // kv descriptor pairs in scan order
+	var rds []proto.RangeDescriptor // corresponding unmarshaled descriptors
 	var intents []proto.Intent
-	var err error
 	if !args.Reverse {
 		// We want to search for the metadata key greater than
 		// args.Key. Scan for both the requested key and the keys immediately
 		// afterwards, up to MaxRanges.
-		startKey, endKey := keys.MetaScanBounds(args.Key)
+		startKey, endKey, err := keys.MetaScanBounds(args.Key)
+		if err != nil {
+			return reply, nil, err
+		}
+
 		// Scan for descriptors.
 		kvs, intents, err = engine.MVCCScan(batch, startKey, endKey, rangeCount,
 			args.Timestamp, consistent, args.Txn)
-	} else {
-		// Use MVCCScan to get first the first range. There are two cases:
-		// 1. args.Key is not an endpoint of the range and
-		// 2. The args.Key is the start/end key of the range.
-		// In the first case, we need use the MVCCScan() to get the first range
-		// descriptor, because the semantic of ReverseScan can't do the work.
-		// For example: If we have ranges ["a","c") and ["c","f") and the
-		// reverse scan request's key range is ["b","d"), then "d".Next() is
-		// less than "f", and so the meta row {"f"->["c", "f")} would be
-		// ignored by MVCCReverseScan. In the second case, the range descriptor
-		// received by MVCCScan will be filtered before results are returned.
-		startKey, endKey := keys.MetaScanBounds(args.Key)
-		kvs, intents, err = engine.MVCCScan(batch, startKey, endKey, 1,
-			args.Timestamp, consistent, args.Txn)
 		if err != nil {
+			// An error here is likely a WriteIntentError when reading consistently.
 			return reply, nil, err
 		}
+	} else {
+		// Use MVCCScan to get first the first range. There are three cases:
+		// 1. args.Key is not an endpoint of the range and
+		// 2a. The args.Key is the start/end key of the range.
+		// 2b. Even worse, the body of args.Key is proto.KeyMax.
+		// In the first case, we need use the MVCCScan() to get the first range
+		// descriptor, because ReverseScan can't do the work.If we have ranges
+		// ["a","c") and ["c","f") and the reverse scan request's key range is
+		// ["b","d"), then "d".Next() is less than "f", and so the meta row
+		// {"f"->["c", "f")} would be ignored by MVCCReverseScan. In case 2a,
+		// the range descriptor received by MVCCScan will be filtered before
+		// results are returned: With ranges ["c", "f") and ["f","z"), reverse
+		// scan on ["d", "f") receives the descriptor {"z"->["f","z")}, which
+		// is discarded below since it's not being asked for. Finally, in case
+		// 2b, we don't even attempt the forward scan because it's neither
+		// defined nor required.
+		// TODO(tschottdorf): unit testing coverage.
+		if args.Key.Less(keys.Meta2KeyMax) {
```

Comment:
ok. Could you add something like:

// Note that Meta1KeyMax is admissible: it means we're looking for the range descriptor that houses `Meta2KeyMax`, and a forward scan handles it correctly.